MAINT: update scipy.special section (#576)
scipy-lectures/scipy-lecture-notes: intro/scipy/index.rst

Special functions: :mod:`scipy.special`
---------------------------------------

> "Special" functions are functions commonly used in science and mathematics that
> are not considered to be "elementary" functions, including

Author: To avoid ambiguity over whether the list is of "special" functions or "elementary" functions:
```suggestion
"Special" functions are functions commonly used in science and mathematics that
are not considered to be "elementary" functions. Examples include
```

 * the gamma function, :func:`scipy.special.gamma`,
 * the error function, :func:`scipy.special.erf`,
 * Bessel functions, such as :func:`scipy.special.jv`
   (Bessel functions of the first kind), and
 * elliptic functions, such as :func:`scipy.special.ellipj`

>    (Jacobian elliptic functions).

Author: ```suggestion
   (Jacobian elliptic functions).

Other special functions are combinations of familiar elementary functions,
but they offer improved accuracy or robustness. For instance, 
:func:`scipy.special.logsumexp([x, y])` computes
:math:`\log(\exp(x) + \exp(y))`, but it is preferable when
the exponential of ``x`` or ``y`` would overflow or underflow.

```

Most of these function are computed elementwise and follow standard
NumPy broadcasting rules when the input arrays have different shapes.
For example, :func:`scipy.special.xlog1py` is mathematically equivalent

> to :math:`x log(1 + y)`.

Author: ```suggestion
to :math:`x\log(1 + y)`.
```

    >>> import scipy as sp
    >>> x = np.asarray([1, 2])
    >>> y = np.asarray([[3], [4], [5]])
    >>> res = sp.special.xlog1py(x, y)
    >>> res.shape
    (2, 3)
    >>> ref = x * np.log(1 + y)
    >>> np.allclose(res, ref)
    True

However, it is numerically favorable for small :math:`y`, when explicit addition
of ``1`` would lead to loss of precision due to floating point truncation error.

    >>> x = 2.5
    >>> y = 1e-18
    >>> x * np.log(1 + y)
    0.0
    >>> sp.special.xlog1py(x, y)
    2.5e-18

> 

Author: This isn't really important to the main ideas, but it seems important to show why such a function is needed.

Many special functions also have "logarithmized" variants. For instance,
the gamma function is related to the factorial function, but it extends
the domain from the positive integers to the complex plane.

   >>> x = np.arange(10)

>    >>> np.allclose(sp.special.gamma(x + 1), sp.special.factorial(x))
>    True

Author: ```suggestion
   >>> np.allclose(sp.special.gamma(x + 1),
   ...             sp.special.factorial(x))
   True
```

   >>> sp.special.gamma(5) < sp.special.gamma(5.5) < sp.special.gamma(6)

>    True

Author: Also not really important to the main idea, but I am reluctant to write more about the `gamma` function without first giving some idea of what it does.

The factorial function grows quickly, and so the gamma function overflows
for moderate values of the argument. However, sometimes only the logarithm
of the gamma function is needed. In such cases, we can compute the logarithm
of the gamma function directly using :func:`scipy.special.gammaln`.

   >>> x = [5, 50, 500]
   >>> np.log(sp.special.gamma(c))
   array([  3.17805383, 144.56574395,          inf])
   >>> sp.special.gammaln(x)
   array([   3.17805383,  144.56574395, 2605.11585036])

> 

Author: I'd love to continue to show the utility of `logsumexp` and such, but maybe that's getting into the weeds. Maybe this is already too much.

Author: ```suggestion
Such functions are often used in combination with 
:func:`scipy.special.logsumexp([x, y])`, which computes
:math:`\log(\exp(x) + \exp(y))` but it is preferable when
the exponential of ``x`` or ``y`` would overflow or underflow.
```

Reviewer: I am interested in seeing what else you would say about `logsumexp` in a follow-up PR.

    >>> np.allclose(res, ref)
    True

However, it is numerically favorable for small :math:`y`, when explicit addition
of ``1`` would lead to loss of precision due to floating point truncation error.

    >>> x = 2.5
    >>> y = 1e-18
    >>> x * np.log(1 + y)
    0.0
    >>> sp.special.xlog1py(x, y)
    2.5e-18

Many special functions also have "logarithmized" variants. For instance,
the gamma function is related to the factorial function, but it extends
the domain from the positive integers to the complex plane.

   >>> x = np.arange(10)
   >>> np.allclose(sp.special.gamma(x + 1), sp.special.factorial(x))
   True
   >>> sp.special.gamma(5) < sp.special.gamma(5.5) < sp.special.gamma(6)
   True

The factorial function grows quickly, and so the gamma function overflows
for moderate values of the argument. However, sometimes only the logarithm
of the gamma function is needed. In such cases, we can compute the logarithm
of the gamma function directly using :func:`scipy.special.gammaln`.

   >>> x = [5, 50, 500]
   >>> np.log(sp.special.gamma(x))
   array([  3.17805383, 144.56574395,          inf])
   >>> sp.special.gammaln(x)
   array([   3.17805383,  144.56574395, 2605.11585036])

> Such functions are often used in combination with 

Author: Not sure why my IDE didn't automatically remove the trailing whitespace.
```suggestion
Such functions are often used in combination with
```
